mvcc/kvstore.go

```diff
@@ -160,6 +166,56 @@ func (s *store) Hash() (hash uint32, revision int64, err error) {
 	return h, s.currentRev, err
 }
 
+func (s *store) HashByRev(rev int64) (hash uint32, currentRev int64, compactRev int64, err error) {
+	readTx := s.b.ReadTx()
+	for {
```

Comment:
@heyitsanthony I wrote a test to illustrate my concern.

```go
func TestHashBeforeCompactionFinishes(t *testing.T) {
	defer testutil.AfterTest(t)

	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
	defer clus.Terminate(t)

	cli.Put(context.TODO(), "foo", "bar2") // rev = 2
	cli.Put(context.TODO(), "foo", "bar3") // rev = 3
	cli.Put(context.TODO(), "bla", "bar4") // rev = 4

	cli := clus.Client(0)
	cli1 := clus.Client(1)

	// compact starting
	cli.Compact(context.TODO(), 3)

	// suppose that ongoing compaction on node 0 hasn't delete anything yet.
	// if HashKV doesn't skip historic keys of "foo" that's before the compact rev 3,, then it will return a hash
	// that's composed of ["foo", "bar2"], ["foo", "bar3"], ["bla", "bar4"].
	// If hashKV skips historic keys of "foo", then it needs to figure out which
	// key rev will be compacted; skip ["foo", "bar2"] since it will be compacted and
	// return hash ["foo", "bar3"] and ["bla", "bar4"].
	resp, _ = toGRPC(cli).Maintainance.HashKV(context.TODO(), &HashKVRequest{Revision: 4})

	// suppose that ongoing compaction on node 1 completes before this Hash begins,
	// then it will return a hash consists only ["foo", "bar3"] and ["bla", "bar4"]
	// regardless skipping historic keys or not.
	resp2, _ = toGRPC(cli2).Maintainance.HashKV(context.TODO(), &HashKVRequest{Revision: 4})

	// expect resp.CompactRevision == resp2.CompactRevision
	// and resp.Hash != resp2.Hash in the scenario no skipping historic keys.
	if resp.CompactRevisions == resp2.CompactRevision {
		if resp.Hash != resp2.CompactRevision {
			t.Fatalf("hash mismatch on same CompactRevision")
		}
	}
}
```